src/main/scala/org/apache/spark/shuffle/ucx/UcxWorkerWrapper.scala

```diff
@@ -0,0 +1,134 @@
+ */
+class UcxWorkerWrapper(worker: UcpWorker, transport: UcxShuffleTransport, ucxConf: UcxShuffleConf,
+                       memoryPool: MemoryPool)
+  extends Closeable with Logging {
+
+  private final val connections = mutable.Map.empty[String, UcpEndpoint]
+  // Tag on which this worker would listen to
+  private val tag = Random.nextLong()
+  private val workerAddress = worker.getAddress
+
+  override def close(): Unit = {
+    connections.foreach{
+      case (_, endpoint) => endpoint.close()
+    }
+    connections.clear()
+    worker.close()
+  }
+
+  /**
+   * The only place for worker progress
+   */
+  private[ucx] def progress() {
+    if ((worker.progress() == 0) && ucxConf.useWakeup) {
+      worker.waitForEvents()
+      worker.progress()
+    }
+  }
+
+  private def getConnection(executorId: String): UcpEndpoint = {
+    val workerAdresses = transport.executorIdToAddress
+
+    if (!workerAdresses.contains(executorId)) {
+      // Block untill there's no worker address for this BlockManagerID
+      val startTime = System.currentTimeMillis()
+      val timeout = ucxConf.getTimeAsMs("spark.network.timeout", "100")
+      workerAdresses.synchronized {
+        while (workerAdresses.get(executorId) == null) {
+          workerAdresses.wait(timeout)
+          if (System.currentTimeMillis() - startTime > timeout) {
+            throw new UcxException(s"Didn't get worker address for $executorId during $timeout")
+          }
+        }
+      }
+    }
+
+    connections.getOrElseUpdate(executorId, {
+      logInfo(s"Worker from thread ${Thread.currentThread().getName} connecting to $executorId")
+      val endpointParams = new UcpEndpointParams()
+        .setUcpAddress(workerAdresses.get(executorId))
+     worker.newEndpoint(endpointParams)
+    })
+  }
+
+  def fetchBlockByBlockId(executorId: String, blockId: BlockId,
+                          resultBuffer: MemoryBlock, cb: OperationCallback): UcxRequest = {
+    val ep = getConnection(executorId)
+    workerAddress.rewind()
+    val mem = memoryPool.get(transport.ucxShuffleConf.rpcMessageSize)
+    val buffer = UcxUtils.getByteBufferView(mem.address,
+      transport.ucxShuffleConf.rpcMessageSize.toInt)
+
+    buffer.putLong(tag)
+    buffer.putInt(workerAddress.capacity())
+    buffer.put(workerAddress)
+
+    val bos = new ByteBufferBackedOutputStream(buffer)
+    val out = new ObjectOutputStream(bos)
+    out.writeObject(blockId)
+    out.flush()
+
+    bos.close()
+
+    logTrace(s"Sending message of size: ${transport.ucxShuffleConf.rpcMessageSize} to fetch $blockId")
+    ep.sendTaggedNonBlocking(mem.address, transport.ucxShuffleConf.rpcMessageSize, tag, new UcxCallback() {
+      override def onSuccess(request: UcpRequest): Unit = {
+        memoryPool.put(mem)
+      }
+    })
+
+    val stats = new UcxStats
+    val result = new UcxSuccessOperationResult(stats)
+    val request = worker.recvTaggedNonBlocking(resultBuffer.address, resultBuffer.size,
```

Comment:
The tag is a single number per thread-local worker as far as I can tell. 

What happens if this worker needs to fetch from two different executors? E.g. `recv(tag, buff1)`, and then `recv(tag, buff2)`. Seems like the order of transmission for `buff1` and `buff2` is dictated by the senders. Shouldn't each "transaction" (`send` + `recv`) have its own tag?

Reply:
This is an implementation of `fetchBlocksByBlockId`. When recv is done your callback would be called. That's how you can distinguish